### DISCUSSION TITLE
Using CPLEx with OSI

### DISCUSSION THREAD

**Discussion by sKamer73** [coin-or/Osi#153] [Q&A]
Hi all,

First thank for your great job! 
I'm contacting you today as I've been trying for the past few hours to use CPLEX from an OsiSolverInterface. 
I've been successfully using CBC with OSI for the past months, but I need to use CPLEX for comparative purposes. I could download and install CPLEx v.20.10 with Academics institutionnal. I am using VS2019 and my OSI version is the one integrated in Cbc-2.10.5.

The issue I'm facing so far is the call of cplex library; I created a new project as libOsiCpx, which compiles correctly. Then in my main project, I added the reference of libOsiCpx, and the cplex library (cplex2010.lib) as Additional Dependencies. If I understood correctly, the displayed error message means that the cplew library is not loaded correctly? 

`1>libOsiCpx.lib(OsiCpxSolverInterface.obj) : error LNK2019: symbole externe non résolu __imp__CPXaddcols@44 référencé dans la fonction "public: virtual void __thiscall OsiCpxSolverInterface::addCol(class CoinPackedVectorBase const &,double,double,double)" (?addCol@OsiCpxSolverInterface@@UAEXABVCoinPackedVectorBase@@NNN@Z)
1>libOsiCpx.lib(OsiCpxSolverInterface.obj) : error LNK2019: symbole externe non résolu __imp__CPXaddfuncdest@16 référencé dans la fonction "private: void __thiscall OsiCpxSolverInterface::gutsOfConstructor(void)" (?gutsOfConstructor@OsiCpxSolverInterface@@AAEXXZ)
1>libOsiCpx.lib(OsiCpxSolverInterface.obj) : error LNK2019: symbole externe non résolu __imp__CPXaddrows@48 référencé dans la fonction "public: virtual void __thiscall OsiCpxSolverInterface::addRow(class CoinPackedVectorBase const &,char,double,double)" (?addRow@OsiCpxSolverInterface@@UAEXABVCoinPackedVectorBase@@DNN@Z)
1>libOsiCpx.lib(OsiCpxSolverInterface.obj) : error LNK2019: symbole externe non résolu __imp__CPXbinvacol@16 référencé dans la fonction "public: virtual void __thiscall OsiCpxSolverInterface::getBInvACol(int,double *)const " (?getBInvACol@OsiCpxSolverInterface@@UBEXHPAN@Z)
`
And it goes on for 65 functions.

As I'm kind of new to C++ and those library considerations, it might be something "classic Cpp" that I just don't do well. I'm as well new to GitHub and all this, so please correct me if I'm not asking at the good place, etc.

If anyone could help me, it would be greatly appreciated.

Cheers!
Simon

**Comment by tkralphs** [comment]
I changed this to a Discussion, since it is better suited as that. GIven the errors you posted above, it seems as though the CPLEX library itself and the header files needed to build `OsiCpx` are being found and the whole thing is building properly, it's just failing at the linking stage because the linker is not finding the symbols corresponding to CPLEX functions inside the library. I'm probably not the right person to give a definitive answer, since I don't really use Visual Studio much. There is probably someone who can say definitively how to do what you're trying to do, but I'll give a few pointers of where to poke around for clues. 

- I believe this can sometimes be caused by incompatible runtimes (see https://docs.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-160), so make sure you know what run-time the CPLEX library is built with. 
- It could also be caused by trying to link a 64-bit library with a 32-bit application (or vice versa). 
- The library itself can be examined with `dumpbin` as a starting point: https://stackoverflow.com/questions/305287/how-to-see-the-contents-of-windows-library-lib). 
- I'm not an expert on exactly how libraries and DLLs work on Windows, but make sure that you are not trying to link to an import library for a DLL rather than a static library.

**Reply by sKamer73** [reply]
Hello,
Thanks a lot for your answer! I've been trying to work on this through your answer (changing architecture, and other parameters) but I couldn't achieve it - still getting the same error message. I'll be trying later, and ask different places. I'll give a feed-back once I'll have the solution! 
I think it might come from my version of Cbc (which I believe is x32), versus my version of Cplex (x64). I never really understood difference between both, but might be the issue!
Thanks,
Simon

**Reply by mjsaltzman** [reply]
That would definitely be an issue. x32 is an old architecture now with 32-bit words and 32-bit addresses. x64 is what all Intel PC and workstation processors are these days, with 64-bit words and addresses. Libraries built for the two different architectures would definitely be incompatible, and the linker building for one of them would likely not even look for libraries built for the other. You should for sure update your Cbc to 64-bit if it isn't already.

**Reply by tkralphs** [reply]
Yeah, you definitely cannot link a 64-bit library to a 32-bit application. I just updated my answer to also point that out. I assume your OS is 64-bit (it would have to be pretty old not to be), so you should be able to just switch your builds to 64-bit within Visual Studio.

**Reply by sKamer73** [reply]
Okay, Thanks for your answer, I think that might be a reason. Nonetheless, I had downloaded source code for CBC and other related Coin-Or projects, and was compiling it under x64 architecture as well as x32 (the famous strategy of try and go ahah). I still had the same issue.

I have something on the fire so I'll be investigating that later on. I was thinking about "re-building" the overall project (i.e. CBC, ..), and I was wondering if you would happen to have some suggestions on this? For a Cbc and OsiCpx project. Last time I used the MSVisual Studio project files, in that one version of Cbc that was well linked to all projects: https://www.coin-or.org/download/source/Cbc/ -> Cbc-2.10.5 (hard to find one version in which Visual Studio does all the work without any issue.. I'm no developper)

**Reply by tkralphs** [reply]
Have you tried the pre-built binaries [here](https://bintray.com/coin-or/download/Cbc/2.10)? That would save you the hassle of building them yourself. The standard build process supported for the Visual Studio compiler is based on using the `coinbrew` script, as documented [here](https://coin-or.github.io/user_introduction). I personally spent far too much time messing around with Visual Studio solution files and gave up supporting them, but there are others who still do use them and I'm sure you'll get plenty of advice if you ask a question over at the Cbc project.

**Reply by sKamer73** [reply]
I did it! I was quite happy I have to say! 
I think the issue was the Win32, x64 issue: my Cbc version was well parametered for x32, but not for x64. Therefore, erasing the libOsiCpx, working out a functioning version of my project in x64, and then add libOsiCpx (with the code from Osi master current branch) did the trick, thanks a lot for your answers!

Quite happy to see this, so I share it:

`CPX0000          Nodes                                         Cuts/
CPX0000     Node  Left     Objective  IInf  Best Integer    Best Bound    ItCnt     Gap

CPX0000  *     0+    0                       1.46078e+09   1.95844e+08            86.59%
CPX0000  *     0+    0                       1.12335e+09   1.95844e+08            82.57%
CPX0000  *     0+    0                       1.11082e+09   1.95844e+08            82.37%
CPX0000        0     0   7.95539e+08     6   1.11082e+09   7.95539e+08      329   28.38%
CPX0000  *     0+    0                       8.01151e+08   7.95539e+08             0.70%
CPX0000        0     0   7.95595e+08     8   8.01151e+08      Fract: 2      339    0.69%
CPX0000  *     0+    0                       7.96027e+08   7.95595e+08             0.05%
CPX0000  *     0+    0                       7.95821e+08   7.95595e+08             0.03%
CPX0000  *     0+    0                       7.95664e+08   7.95595e+08             0.01%
CPX0000
CPX0000  Gomory fractional cuts applied:  2`

**Reply by jhmgoossens** [reply]
>messing around with Visual Studio solution files

Indeed, several of the projects in cbc.sln for VS2019 had issues building due to contradicting project settings. I created PR #155 for Osi and PR [#372 in Cbc](https://github.com/coin-or/Cbc/pull/372) to try to correct. Notably the Release / x64 was not building when using the VS2019 solution.
